server/src/main/webapp/WEB-INF/rails/webpack/views/pages/new-environments/create_env_modal.tsx

```diff
@@ -0,0 +1,102 @@
+import m from "mithril";
+import Stream from "mithril/stream";
+import {EnvironmentVariablesWithOrigin} from "models/new-environments/environment_environment_variables";
+import {Pipelines} from "models/new-environments/environment_pipelines";
+import {Environments, EnvironmentWithOrigin} from "models/new-environments/environments";
+import {EnvironmentsAPIs} from "models/new-environments/environments_apis";
+import {Origin, OriginType} from "models/new-environments/origin";
+import {ButtonGroup, Cancel, Primary} from "views/components/buttons";
+import {FlashMessage, MessageType} from "views/components/flash_message";
+import {Form} from "views/components/forms/form";
+import {TextField} from "views/components/forms/input_fields";
+import {Modal} from "views/components/modal";
+
+export class CreateEnvModal extends Modal {
+  private readonly onSuccessfulSave: (msg: m.Children) => void;
+  private environment: EnvironmentWithOrigin;
+  private environments: Stream<Environments>;
+  private errorMessage: Stream<string> = Stream();
+
+  constructor(environments: Stream<Environments>, onSuccessfulSave: (msg: m.Children) => void) {
+    super();
+    this.environments     = environments;
+    this.onSuccessfulSave = onSuccessfulSave;
+    this.environment      = new EnvironmentWithOrigin("",
+                                                      [new Origin(OriginType.GoCD)],
+                                                      [],
+                                                      new Pipelines(),
+                                                      new EnvironmentVariablesWithOrigin());
+  }
+
+  body(): m.Children {
+    let flashMsg;
+    if (this.errorMessage()) {
+      flashMsg = <FlashMessage type={MessageType.alert} message={this.errorMessage()}/>;
+    }
+    return <Form>
+      <FlashMessage type={MessageType.info}
+                    message={"Pipelines, agents and environment variables can be added post creation of the environment."}/>
+      {flashMsg}
+      <TextField label="Environment name"
+                 property={this.environment.name}
+                 errorText={this.environment.errors().errorsForDisplay("name")}
+                 required={true}/>
+    </Form>;
+  }
+
+  title(): string {
+    return "Add New Environment";
+  }
+
+  buttons() {
+    return [<ButtonGroup>
+      <Cancel data-test-id="button-cancel" onclick={() => this.close()}>Cancel</Cancel>
+      <Primary data-test-id="button-save" onclick={this.validateAndPerformSave.bind(this)}>Save</Primary>
+    </ButtonGroup>];
+  }
+
+  private validateAndPerformSave() {
+    if (!this.environment.isValid()) {
+      return;
+    }
+    EnvironmentsAPIs.create(this.environment)
+                   .then((result) => {
+                     result.do(
+                       (successResponse: SuccessResponse<ObjectWithEtag<EnvironmentWithOrigin>>) => {
+                         this.environments().push(successResponse.body.object);
```

Comment:
I don't think there is a need to add the environment here as we fetch data on successful save. Can you please check once?

Reply:
No we don't fetch the data on `onSuccessfulSave`. We just update the flash message.